pkg/parse/parser.go

```diff
@@ -57,86 +49,87 @@ func (f AnyFormat) String() string {
-	details.bpCountTOML = countSetFieldsRecursive(bpTOML)
+		return ubpData, nil
+	} else {
+		errs = append(errs, fmt.Errorf("%w (UBP-JSON)", err))
+	}
 
 	// Try BP JSON
-	bpTempJSON := new(bp.Blueprint)
-	bpErrJSON := json.Unmarshal(buf, bpTempJSON)
-	details.bpCountTemp = countSetFieldsRecursive(bpErrJSON)
-	imJSON := conv.NewInternalImporter(bpTempJSON)
-	bpJSON, bpWarnJSON := imJSON.Import()
-	details.bpCountJSON = countSetFieldsRecursive(bpJSON)
-
-	maxCount := max(details.ubpCountYAML, details.ubpCountJSON, details.bpCountTOML, details.bpCountJSON)
-	err := errors.Join(
-		fmt.Errorf("YAML: %w", ubpErrYAML),
-		fmt.Errorf("JSON: %w", ubpErrJSON),
-		fmt.Errorf("TOML: %w", bpErrTOML),
-		fmt.Errorf("JSON: %w", bpErrJSON),
-	)
-	details.Warnings = errors.Join(bpWarnTOML, bpWarnJSON)
-
-	errDefaults := ubp.PopulateDefaults(bpTOML)
-	if errDefaults != nil {
-		err = errors.Join(err, fmt.Errorf("populate defaults for BP-TOML: %w", errDefaults))
-	}
+	bpData := new(bp.Blueprint)
+	dec := json.NewDecoder(bytes.NewReader(buf))
+	dec.DisallowUnknownFields()
+	err = dec.Decode(bpData)
+	if err == nil {
+		details.Format = AnyFormatBPJSON
+
+		importer := conv.NewInternalImporter(bpData)
+		imData, warn := importer.Import()
+		details.Intermediate = bpData
+		details.Warnings = warn
 
-	errDefaults = ubp.PopulateDefaults(bpJSON)
-	if errDefaults != nil {
-		err = errors.Join(err, fmt.Errorf("populate defaults for BP-JSON: %w", errDefaults))
+		errDefaults := ubp.PopulateDefaults(imData)
+		return imData, errDefaults
+	} else {
+		errs = append(errs, fmt.Errorf("%w (BP-JSON)", err))
 	}
 
-	if ubpErrYAML == nil && details.ubpCountYAML == maxCount {
-		details.Format = AnyFormatUBPYAML
-		return ubpYAML, nil
-	} else if ubpErrJSON == nil && details.ubpCountJSON == maxCount {
-		details.Format = AnyFormatUBPJSON
-		return ubpJSON, nil
-	} else if bpErrTOML == nil && details.bpCountTOML == maxCount {
+	// Try BP TOML (this parser does not have a strict mode, so it comes last)
+	bpData = new(bp.Blueprint)
+	err = toml.Unmarshal(buf, bpData)
+	if err == nil {
 		details.Format = AnyFormatBPTOML
-		details.Converted = true
-		details.Intermediate = bpTempTOML
-		return bpTOML, nil
-	} else if bpErrJSON == nil && details.bpCountJSON == maxCount {
-		details.Format = AnyFormatBPJSON
-		details.Converted = true
-		details.Intermediate = bpTempJSON
-		return bpJSON, nil
+
+		importer := conv.NewInternalImporter(bpData)
+		imData, warn := importer.Import()
+		details.Intermediate = bpData
+		details.Warnings = warn
+
+		errDefaults := ubp.PopulateDefaults(imData)
+		return imData, errDefaults
 	} else {
-		countErr := fmt.Errorf("fields set: UBPY:%d UBPJ:%d BPT:%d BPJ:%d",
-			details.ubpCountYAML,
-			details.ubpCountJSON,
-			details.bpCountTOML,
-			details.bpCountJSON,
-		)
-		return nil, errors.Join(ErrParsingExplanation, err, countErr)
+		errs = append(errs, fmt.Errorf("%w (BP-TOML)", err))
 	}
+
+	return nil, errors.Join(errs...)
 }
```

Comment:
This is the main gist of the patch, instead of counting via reflection it does what is obvious now. I had no idea strict mode exists, or at least I have not realized this is the best way to solve this.